pkg/kubectl/cmd/expose.go

```diff
@@ -147,101 +137,124 @@ func RunExpose(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []str
 		return cmdutil.UsageError(cmd, fmt.Sprintf("generator %q not found.", generatorName))
 	}
 	names := generator.ParamNames()
-	params := kubectl.MakeParams(cmd, names)
-	name := info.Name
-	if len(name) > validation.DNS952LabelMaxLength {
-		name = name[:validation.DNS952LabelMaxLength]
+
+	// Get the input object
+	inputObject, err := r.Object()
+	if err != nil {
+		return err
 	}
-	params["default-name"] = name
 
-	// For objects that need a pod selector, derive it from the exposed object in case a user
-	// didn't explicitly specify one via --selector
-	if s, found := params["selector"]; found && kubectl.IsZero(s) {
-		s, err := f.MapBasedSelectorForObject(inputObject)
-		if err != nil {
-			return cmdutil.UsageError(cmd, fmt.Sprintf("couldn't retrieve selectors via --selector flag or introspection: %s", err))
-		}
-		params["selector"] = s
+	var inputObjects []runtime.Object
+	switch inputObject.(type) {
+	case *api.List:
+		inputObjects = inputObject.(*api.List).Items
+	default:
+		inputObjects = append(inputObjects, inputObject)
 	}
 
-	// For objects that need a port, derive it from the exposed object in case a user
-	// didn't explicitly specify one via --port
-	if port, found := params["port"]; found && kubectl.IsZero(port) {
-		ports, err := f.PortsForObject(inputObject)
-		if err != nil {
-			return cmdutil.UsageError(cmd, fmt.Sprintf("couldn't find port via --port flag or introspection: %s", err))
+	for i := range infos {
```

Comment:
I like this in principle, but I'd like to do it in a separate pull.  `--recursive` still makes sense for single resource commands, since I can point to a parent dir and this will navigate and find my single file at a lower level (its really the same argument as allowing directory references).